### ISSUE TITLE
Proposal: Disallow setting a custom EventEmitter, allow to set a custom reporter

### ISSUE THREAD

**Issue by honzajavorek** [apiaryio/dredd#1428]
**Is your feature request related to a problem? Please describe.**

It is hard to test Dredd.

**Describe the solution you'd like**

Currently I can provide my own `EventEmitter` instance, but perhaps it would make more sense if I could provide a custom reporter, which gets all the data anyway - so there could be a reporter set by tests, which captures whatever needs to be asserted. Passing custom `EventEmitter` in configuration feels weird to me, even for the purpose of testing.

Does `EventEmitter` have any other purpose than testing? Are people using it? 🤔 

**Describe alternatives you've considered**

See https://github.com/apiaryio/dredd/issues/341, https://github.com/apiaryio/dredd/issues/922

**Comment by artem-zakharchenko** [comment]
I agree. Using a custom `EventEmitter` is not only confusing, but tempting to assert Dredd internals. I believe we are speaking about integration tests now, where the public info is what should be asserted.

For example, instead of provisioning a custom emitter here:

https://github.com/apiaryio/dredd/blob/cd9538f482f00926a3642f71eeb03af81bcaf9ef/test/integration/sanitation-test.js#L15-L35

We could assert this info from the Dredd's output (i.e. stats). My philosophy would be: what is not in the public output doesn't need an assertion in the integration tests. It may be something to cover in unit tests, or remove at all.

Can we use Dredd's stats for this kind of assertions?

**Comment by honzajavorek** [comment]
We could, yes. But turning stats into a an object which is able to contain all information Dredd produces during testing is almost the same task as providing the information in a machine-readable reporter.

We can design the structure once and use it consistently in both stats and as an output of the reporter. The difference being, reporter is incremental (prints to output _during_ testing), CLI-user-oriented, and end-to-end - while the stats are available only once testing is done (`const stats = await dredd()`) and they're JS-user-oriented and powering the integration tests.

Having it in stats instead of the `EventEmitter` removes the ability for users to inspect Dredd testing _while_ it happens, but I don't know if there is a real user value behind that.
